### PR TITLE
#5 Preprocessing pipelines for Text and Audio

### PR DESCRIPTION
**Pull request by botirk38**
## Why?

We need to implement these preprocessing pipelines to standardize and enhance our data processing workflow for both text and audio data. These pipelines will:

1. Improve text data quality by handling missing values and performing sentence segmentation.
2. Prepare audio data for further processing by extracting features using the Whisper model.

These preprocessing steps are crucial for downstream tasks such as natural language processing and speech recognition.

## How?

We've implemented two main preprocessing pipelines:

1. TextPreprocessingPipeline:
   - Handles null and missing values based on configuration (skip, remove, or fill).
   - Performs sentence segmentation using spaCy models.
   - Supports multiple languages through the SPACY_MODELS dictionary.

2. AudioPreprocessingPipeline:
   - Uses Whisper's feature extractor to preprocess audio data.
   - Supports limiting audio duration.
   - Handles different input formats (numpy array or dictionary with array and sampling rate).

Technical decisions:
- Used dataclasses for configuration to ensure type safety and ease of use.
- Implemented abstract base classes (ABC) to define a common interface for preprocessing pipelines.
- Utilized spaCy for text preprocessing due to its efficiency and multi-language support.
- Chose Whisper for audio preprocessing as it's a state-of-the-art model for speech processing tasks.

WIP: 
- Error handling for edge cases in audio preprocessing.
- Optimization for large-scale data processing.

## Test plan

To test these changes:

1. Text Preprocessing:
   ```python
   from huggingface_pipelines.preprocessing import TextPreprocessingPipeline, TextPreprocessingPipelineConfig
   
   config = TextPreprocessingPipelineConfig(
       columns=["text"],
       handle_missing="skip",
       source_lang="eng_Latn"
   )
   pipeline = TextPreprocessingPipeline(config)
   
   test_batch = {
       "text": ["This is a test.", "Another sentence. And one more.", None, ""]
   }
   result = pipeline.process_batch(test_batch)
   print(result)

2. Audio Preprocessing
      ```python
      import numpy as np
      from huggingface_pipelines.preprocessing import AudioPreprocessingPipeline, AudioPreprocessingPipelineConfig
      
      config = AudioPreprocessingPipelineConfig(
          columns=["audio"],
          whisper_model="openai/whisper-base",
          max_duration=30.0
      )
      pipeline = AudioPreprocessingPipeline(config)
      
      # Generate dummy audio data
      sample_rate = 16000
      dummy_audio = np.random.randn(sample_rate * 10)  # 10 seconds of audio
      
      test_batch = {
          "audio": [{"array": dummy_audio, "sampling_rate": sample_rate}]
      }
      result = pipeline.process_batch(test_batch)
      print(result["audio_preprocessed"][0].shape)
      
      ```

### CODE REVIEW

File: huggingface_pipelines/preprocessing.py [hunk 1]
```diff
@@ -0,0 +1,201 @@
+from transformers import WhisperProcessor, WhisperFeatureExtractor
+import numpy as np
+import spacy
```
Comment (artemru) [inline huggingface_pipelines/preprocessing.py:3]: you need to add spacy as depedancy at pyproject file. Same for transformers I suppose.

File: huggingface_pipelines/preprocessing.py [hunk 2]
```diff
@@ -0,0 +1,201 @@
...
+        """
+        for column in self.config.columns:
+            if column in batch:
+                batch[f"{column}_preprocessed"] = [
+                    self.preprocess_text(text) for text in batch[column]]
+
+                logger.info(
+                    f"Preprocessed {column} column:  {batch[column][:5]}")
+            else:
+                logger.warning(f"Column {column} not found in batch.")
```
Comment (artemru) [inline huggingface_pipelines/preprocessing.py:60]: I dont think we want to see all this logs... there'll be a lot of them in normal case and they will not be informative.

File: huggingface_pipelines/preprocessing.py [hunk 3]
```diff
@@ -0,0 +1,201 @@
...
+        self.nlp = self.load_spacy_model(self.config.source_lang)
+        logger.info("Model initialized.")
+
+    def load_spacy_model(self, lang_code: str):
+        """
+        Loads the appropriate spaCy model based on the language code.
+        """
+        if lang_code not in SPACY_MODELS:
+            raise ValueError(
+                f"Unsupported language code: {lang_code}. Please add it to the SPACY_MODELS dictionary.")
```
Comment (artemru) [inline huggingface_pipelines/preprocessing.py:93]: 👍

**Review by antoine-tran — COMMENTED** [review]
About the premise of this PR:

- The text bit: It seems to me we want to develop not a general text processing pipeline, but one particular kind (Sentence segmentation), that might be using spacy or other sentencizer model. 
- The audio bit: We want to extract features from Whisper, do we ensure these features are compatible with other non-Whisper models ? If not, what is the use case of the pipeline

Also, in both cases, I was wondering if making the features as complete pipelines is the best solution. If we want to streamline these with the instead of a function that can be called optionally to transform  the HG dataset within some other pipeline, such as e.g. `TextToEmnbedingPipeline` or `SpeechToEmbeddingPipeline`
